src/Microsoft.DotNet.ProjectModel/Files/IncludeFilesResolver.cs

```diff
@@ -0,0 +1,182 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.DotNet.ProjectModel.Utilities;
+using Microsoft.Extensions.FileSystemGlobbing;
+
+namespace Microsoft.DotNet.ProjectModel.Files
+{
+    public class IncludeFilesResolver
+    {
+        public static IEnumerable<IncludeEntry> GetIncludeFiles(IncludeContext context, string targetBasePath, IList<DiagnosticMessage> diagnostics)
+        {
+            return GetIncludeFiles(context, targetBasePath, diagnostics, flatten: false);
+        }
+
+        public static IEnumerable<IncludeEntry> GetIncludeFiles(
+            IncludeContext context,
+            string targetBasePath,
+            IList<DiagnosticMessage> diagnostics,
+            bool flatten)
+        {
+            var sourceBasePath = PathUtility.EnsureTrailingSlash(context.SourceBasePath);
+            targetBasePath = PathUtility.GetPathWithDirectorySeparator(targetBasePath);
+
+            var includeEntries = new HashSet<IncludeEntry>();
+
+            // Check for illegal characters in target path
+            if (string.IsNullOrEmpty(targetBasePath))
+            {
+                diagnostics?.Add(new DiagnosticMessage(
+                    ErrorCodes.NU1003,
+                    $"Invalid '{context.Option}' section. The target '{targetBasePath}' is invalid, " +
+                    "targets must either be a file name or a directory suffixed with '/'. " +
+                    "The root directory of the package can be specified by using a single '/' character.",
+                    sourceBasePath,
+                    DiagnosticMessageSeverity.Error));
+            }
+            else if (targetBasePath.Split('/').Any(s => s.Equals(".") || s.Equals("..")))
+            {
+                diagnostics?.Add(new DiagnosticMessage(
+                    ErrorCodes.NU1004,
+                    $"Invalid '{context.Option}' section. " +
+                    $"The target '{targetBasePath}' contains path-traversal characters ('.' or '..'). " +
+                    "These characters are not permitted in target paths.",
+                    sourceBasePath,
+                    DiagnosticMessageSeverity.Error));
+            }
+            else
+            {
+                var files = GetIncludeFilesCore(
+                    sourceBasePath,
+                    context.IncludePatterns,
+                    context.ExcludePatterns,
+                    context.IncludeFiles,
+                    context.ExcludeFiles,
+                    context.BuiltInsInclude,
+                    context.BuiltInsExclude);
+
+                var isFile = !targetBasePath.EndsWith(Path.DirectorySeparatorChar.ToString());
```

Comment:
can't you just compare the last character directly instead of create another string?